File: backend/src/main/java/com/hostel/backend/scheduler/SlaScheduler.java
package com.hostel.backend.scheduler;

import com.hostel.backend.entity.Complaint;
import com.hostel.backend.entity.User;
import com.hostel.backend.enums.ComplaintPriority;
import com.hostel.backend.enums.ComplaintStatus;
import com.hostel.backend.enums.Role;
import com.hostel.backend.repository.ComplaintRepository;
import com.hostel.backend.repository.UserRepository;
import com.hostel.backend.service.NotificationService;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Component
public class SlaScheduler {

    private final ComplaintRepository complaintRepository;
    private final UserRepository userRepository;
    private final NotificationService notificationService;

    public SlaScheduler(ComplaintRepository complaintRepository,
                        UserRepository userRepository,
                        NotificationService notificationService) {
        this.complaintRepository = complaintRepository;
        this.userRepository      = userRepository;
        this.notificationService = notificationService;
    }

    // Runs every hour
    @Scheduled(fixedRate = 3_600_000)
    @Transactional
    public void checkSlaBreaches() {
        LocalDateTime now          = LocalDateTime.now();
        LocalDateTime overdueCutoff    = now.minusDays(7);
        LocalDateTime escalationCutoff = now.minusHours(24);

        List<User> staff = userRepository.findByRole(Role.WARDEN);
        staff.addAll(userRepository.findByRole(Role.CARETAKER));

        List<ComplaintStatus> activeStatuses = List.of(
                ComplaintStatus.CREATED,
                ComplaintStatus.ASSIGNED,
                ComplaintStatus.IN_PROGRESS);

        List<Complaint> activeComplaints =
                complaintRepository.findByStatusIn(activeStatuses);

        for (Complaint c : activeComplaints) {

            // ── S5: Mark OVERDUE if older than 7 days ──────────────────────
            if (!c.isOverdue() && c.getCreatedAt().isBefore(overdueCutoff)) {
                c.setOverdue(true);
                c.setLastSlaNotifiedAt(now);
                complaintRepository.save(c);

                String msg = "🔴 Complaint #" + c.getId() + " '" + c.getTitle()
                        + "' is OVERDUE (>7 days unresolved)";
                staff.forEach(s -> notificationService.sendNotification(s, msg, "SLA_OVERDUE"));
                notificationService.sendNotification(c.getStudent(), msg, "SLA_OVERDUE");
            }

            // ── S10: Escalate HIGH priority not assigned within 24hrs ───────
            if (!c.isEscalated()
                    && c.getPriority() == ComplaintPriority.HIGH
                    && c.getStatus() == ComplaintStatus.CREATED
                    && c.getCreatedAt().isBefore(escalationCutoff)) {
                c.setEscalated(true);
                c.setLastSlaNotifiedAt(now);
                complaintRepository.save(c);

                String msg = "🚨 HIGH priority complaint #" + c.getId()
                        + " '" + c.getTitle() + "' is unassigned for >24 hours!";
                staff.forEach(s -> notificationService.sendNotification(s, msg, "ESCALATED"));
            }
        }
    }
}

File: backend/src/jobs/slaScheduler.js
const cron = require('node-cron');
const prisma = require('../config/prisma');
const logger = require('../config/logger');
const notificationService = require('../services/notification.service');

const ACTIVE_STATUSES = ['CREATED', 'ASSIGNED', 'IN_PROGRESS'];

// Mirrors scheduler/SlaScheduler.java exactly: runs every hour, flags
// complaints unresolved >7 days as overdue, and unassigned HIGH-priority
// complaints older than 24h as escalated.
async function checkSlaBreaches() {
  const now = new Date();
  const overdueCutoff = new Date(now.getTime() - 7 * 24 * 60 * 60 * 1000);
  const escalationCutoff = new Date(now.getTime() - 24 * 60 * 60 * 1000);

  const staff = await prisma.user.findMany({ where: { role: { in: ['WARDEN', 'CARETAKER'] } } });
  const activeComplaints = await prisma.complaint.findMany({
    where: { status: { in: ACTIVE_STATUSES } },
    include: { student: true },
  });

  for (const c of activeComplaints) {
    // S5: mark OVERDUE if older than 7 days
    if (!c.isOverdue && c.createdAt && c.createdAt < overdueCutoff) {
      await prisma.complaint.update({
        where: { id: c.id },
        data: { isOverdue: true, lastSlaNotifiedAt: now },
      });

      const msg = `🔴 Complaint #${c.id} '${c.title}' is OVERDUE (>7 days unresolved)`;
      await Promise.all(staff.map((s) => notificationService.sendNotification(s, msg, 'SLA_OVERDUE')));
      await notificationService.sendNotification(c.student, msg, 'SLA_OVERDUE');
    }

    // S10: escalate HIGH priority not assigned within 24h
    if (!c.isEscalated && c.priority === 'HIGH' && c.status === 'CREATED' && c.createdAt && c.createdAt < escalationCutoff) {
      await prisma.complaint.update({
        where: { id: c.id },
        data: { isEscalated: true, lastSlaNotifiedAt: now },
      });

      const msg = `🚨 HIGH priority complaint #${c.id} '${c.title}' is unassigned for >24 hours!`;
      await Promise.all(staff.map((s) => notificationService.sendNotification(s, msg, 'ESCALATED')));
    }
  }
}

function startSlaScheduler() {
  // Original: @Scheduled(fixedRate = 3_600_000) — every hour, on the hour.
  cron.schedule('0 * * * *', () => {
    checkSlaBreaches().catch((err) => logger.error('SLA scheduler run failed', err));
  });
  logger.info('SLA scheduler started (runs hourly)');
}

module.exports = { startSlaScheduler, checkSlaBreaches };
